ml/backend.go

```diff
@@ -100,8 +101,17 @@ type Context interface {
 
 	Forward(...Tensor) Context
 	Compute(...Tensor)
-	MaxTensors() int
+	MaxGraphNodes() int
 	Close()
+
+	// Input returns a context appropriate for creating input tensors
+	Input() Context
+
+	// Output returns a context appropriate for creating output tensors
+	Output() Context
+
+	// Layer returns a context appropriate for creating intermediate tensors
+	Layer(int) Context
```

Comment:
Clients always need to call one of these function, right? What happens if they don't and is there a way to avoid surprises, such as poor performance, if they use the bare context?

Reply:
That's a good point. I was thinking of adding a new type for creating tensors but decided not to to reduce adding yet another interface